File: runelite-client/src/main/java/net/runelite/client/plugins/alfred/scripts/transporthelper/TransportHelperPlugin.java
package net.runelite.client.plugins.alfred.scripts.transporthelper;

import lombok.extern.slf4j.Slf4j;
import net.runelite.api.Client;
import net.runelite.api.GameObject;
import net.runelite.api.Perspective;
import net.runelite.api.Tile;
import net.runelite.api.coords.LocalPoint;
import net.runelite.api.coords.WorldPoint;
import net.runelite.client.plugins.Plugin;
import net.runelite.client.plugins.PluginDescriptor;
import net.runelite.client.plugins.alfred.Alfred;
import net.runelite.client.plugins.alfred.api.rs.walk.RSTile;
import net.runelite.client.ui.overlay.OverlayManager;

import javax.inject.Inject;
import java.awt.*;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;

@PluginDescriptor(name = "Alfred Transport Helper", enabledByDefault = false)
@Slf4j
public class TransportHelperPlugin extends Plugin {
    @Inject
    private OverlayManager overlayManager;
    @Inject
    private TransportHelperOverlay overlay;
    private MouseListener mouseListener;

    @Override
    protected void startUp() throws Exception {
        overlayManager.add(overlay);
        setupMouseListener();
    }

    @Override
    protected void shutDown() throws Exception {
        overlayManager.remove(overlay);
        Alfred.getClient().getCanvas().removeMouseListener(mouseListener);
    }

    private void setupMouseListener() {
        mouseListener = new MouseListener() {
            @Override
            public void mouseClicked(MouseEvent e) {
                Client client = Alfred.getClient();
                for (RSTile tile : Alfred.api.walk().getAllTiles()) {
                    final LocalPoint tileLocalLocation = tile.getTile().getLocalLocation();
                    Polygon poly = Perspective.getCanvasTilePoly(client, tileLocalLocation);
                    if (poly != null && poly.contains(client.getMouseCanvasPosition().getX(), client.getMouseCanvasPosition().getY())) {
                        getTileTransportInformation(tile.getTile());
                    }
                }
            }

            @Override
            public void mousePressed(MouseEvent e) {

            }

            @Override
            public void mouseReleased(MouseEvent e) {

            }

            @Override
            public void mouseEntered(MouseEvent e) {

            }

            @Override
            public void mouseExited(MouseEvent e) {

            }
        };

        Alfred.getClient().getCanvas().addMouseListener(mouseListener);
    }

    private void getTileTransportInformation(Tile tile) {
        String objectName = "";
        int objectId = -1;

        WorldPoint worldPoint = tile.getWorldLocation();


        if (tile.getWallObject() != null) {
            objectName = Alfred.api.objects().getObjectIdVariableName(tile.getWallObject().getId());
            objectId = tile.getWallObject().getId();
        } else {
            for (GameObject gameObject : tile.getGameObjects()) {
                if (gameObject == null) {
                    continue;
                }

                objectName = Alfred.api.objects().getObjectIdVariableName(gameObject.getId());
                objectId = gameObject.getId();
                break;
            }
        }

        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("(");

        // Start tile
        stringBuilder.append(String.format("%d, %d, %d", worldPoint.getX(), worldPoint.getY(), worldPoint.getPlane()));
        stringBuilder.append(", ");

        // End tile
        stringBuilder.append("None, None, None");
        stringBuilder.append(", ");

        // Name
        stringBuilder.append("'Name'");
        stringBuilder.append(", ");

        // Object ID
        stringBuilder.append(String.format("%d", objectId));
        stringBuilder.append(", ");

        // Object Name
        stringBuilder.append(String.format("'%s'", objectName));
        stringBuilder.append("),");

        System.out.println(stringBuilder.toString());
    }
}

File: runelite-client/src/main/java/net/runelite/client/plugins/alfred/scripts/transporthelper/TransportHelperPlugin.kt
package net.runelite.client.plugins.alfred.scripts.transporthelper

import com.google.gson.GsonBuilder
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.google.inject.Provides
import net.runelite.api.Perspective
import net.runelite.api.Tile
import net.runelite.client.config.ConfigManager
import net.runelite.client.plugins.Plugin
import net.runelite.client.plugins.PluginDescriptor
import net.runelite.client.plugins.alfred.Alfred
import net.runelite.client.plugins.alfred.enums.TransportTypes
import net.runelite.client.ui.overlay.OverlayManager
import java.awt.Toolkit
import java.awt.datatransfer.StringSelection
import java.awt.event.MouseEvent
import java.awt.event.MouseListener
import javax.inject.Inject

@PluginDescriptor(name = TransportHelperPlugin.CONFIG_GROUP, enabledByDefault = false)
class TransportHelperPlugin : Plugin() {
    companion object {
        const val CONFIG_GROUP = "Alfred Transport Helper"
    }

    @Inject
    private lateinit var config: TransportHelperConfig

    @Provides
    fun provideConfig(configManager: ConfigManager): TransportHelperConfig {
        return configManager.getConfig(TransportHelperConfig::class.java)
    }

    @Inject
    private lateinit var overlayManager: OverlayManager

    @Inject
    private lateinit var overlay: TransportHelperOverlay
    private var mouseListener: MouseListener? = null

    override fun startUp() {
        overlayManager.add(overlay)
        setupMouseListener()
    }

    override fun shutDown() {
        overlayManager.remove(overlay)
        Alfred.client.getCanvas().removeMouseListener(mouseListener)
    }

    private fun setupMouseListener() {
        mouseListener = object : MouseListener {
            override fun mouseClicked(e: MouseEvent) {
                val client = Alfred.client
                for (tile in Alfred.api.walk.getAllTiles()) {
                    val tileLocalLocation = tile.tile.getLocalLocation()
                    val poly = Perspective.getCanvasTilePoly(client, tileLocalLocation)
                    if (poly != null && poly.contains(client.getMouseCanvasPosition().x, client.getMouseCanvasPosition().y)) {
                        getTileTransportInformation(tile.tile)
                    }
                }
            }

            override fun mousePressed(e: MouseEvent) {}
            override fun mouseReleased(e: MouseEvent) {}
            override fun mouseEntered(e: MouseEvent) {}
            override fun mouseExited(e: MouseEvent) {}
        }
        Alfred.client.getCanvas().addMouseListener(mouseListener)
    }

    private fun getTileTransportInformation(tile: Tile) {
        var objectName: String? = ""
        var objectId = -1
        var objectHash = -1
        val worldPoint = tile.worldLocation
        var eastWest = false

        if (tile.wallObject != null) {
            objectName = Alfred.api.objects.getObjectIdVariableName(tile.wallObject.id)
            objectId = tile.wallObject.getId()
            objectHash = tile.wallObject.hash.toInt()
            if (tile.wallObject.orientationA == 1 || tile.wallObject.orientationA == 4) {
                eastWest = true
            }
        } else {
            for (gameObject in tile.getGameObjects()) {
                if (gameObject == null) {
                    continue
                }
                objectName = Alfred.api.objects.getObjectIdVariableName(gameObject.id)
                objectId = gameObject.getId()
                objectHash = gameObject.hash.toInt()
                println("Game Object Orientation: ${gameObject.orientation}")
                println("Game Object Model Orientation: ${gameObject.modelOrientation}")
                break
            }
        }

        val transport = JsonObject()
        transport.addProperty("transport_name", "")
        transport.addProperty("object_hash", objectHash)
        transport.addProperty("object_id", objectId)
        transport.addProperty("object_name", objectName)

        if (config.transportType() == TransportTypes.DOOR) {
            if (eastWest) {
                transport.addProperty("unblock_north_south", false)
                transport.addProperty("unblock_east_west", true)
            } else {
                transport.addProperty("unblock_north_south", true)
                transport.addProperty("unblock_east_west", false)
            }
        } else {
            transport.addProperty("unblock_north_south", false)
            transport.addProperty("unblock_east_west", false)
        }

        val startTile = JsonObject()
        startTile.addProperty("x", worldPoint.x)
        startTile.addProperty("y", worldPoint.y)
        startTile.addProperty("z", worldPoint.plane)

        val endTile = JsonObject()
        endTile.addProperty("x", 0)
        endTile.addProperty("y", 0)
        endTile.addProperty("z", 0)

        val connection = JsonObject()
        connection.add("start_tile", startTile)
        connection.add("end_tile", endTile)
        connection.addProperty("action", "")

        val connections = JsonArray()
        connections.add(connection)

        transport.add("connections", connections)

        val gson = GsonBuilder().setPrettyPrinting().create()

        val stringSelection = StringSelection(gson.toJson(transport))
        val clipboard = Toolkit.getDefaultToolkit().systemClipboard
        clipboard.setContents(stringSelection, null)
        println(gson.toJson(transport))

    }
}
